document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -82,6 +84,68 @@ public boolean upsert(Key key, Document document) throws IOException {
     }
   }
 
+  /**
+   * Update an existing document if condition is evaluated to true. Conditional will help in
+   * providing optimistic locking support for concurrency update.
+   */
+  @Override
+  public DocStoreResult update(Key key, Document document, Filter condition) throws IOException {
+    StringBuilder upsertQueryBuilder = new StringBuilder(getUpdateSQL());
+
+    String jsonString = prepareDocument(key, document);
+    Params.Builder paramsBuilder = Params.newBuilder();
+    paramsBuilder.addObjectParam(key.toString());
+    paramsBuilder.addObjectParam(jsonString);
+
+    if (condition != null) {
+      String conditionQuery = parseFilter(condition, paramsBuilder);
+      if (conditionQuery != null) {
+        upsertQueryBuilder.append(" WHERE ").append(conditionQuery);
+      }
+    }
+
+    try (PreparedStatement preparedStatement =
+        buildPreparedStatement(upsertQueryBuilder.toString(), paramsBuilder.build())) {
+      int result = preparedStatement.executeUpdate();
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Write result: {}", result);
+      }
+      DocStoreResult.Builder resultBuilder = DocStoreResult.Builder.newBuilder();
+      resultBuilder.setSuccess(result > 0).setResultCount(result);
+      if (result <= 0) {
```

Comment:
I'm not sure if result=0 can be considered a failure. Isn't it something for the consumers/clients to decide based on the business logic/use case

Reply:
On condition fail, and the document is not updated, it is not an error. So, the API will not throw any error. And as per the java doc of prepared stmt, it returns `0` if there is no modification.

> @return either (1) the row count for SQL Data Manipulation Language (DML) statements or (2) 0 for SQL statements that return nothing

So, if we keep the `success` status like all othersAPIs, we have to determine using `result`. If we don't we can only keep `resultCount` in DocStoreResult.

Reply:
As discussed with @avinashkolluru, we will return the following response object and client will make a call accordingly.
For create API:
```
public class CreateResult {
  private long createdCount;
}
```
For Update API
```
public class UpdateResult {
  private long updatedCount;
}
```

Reply:
@avinashkolluru I have addressed this.